cc/testing/evaluation/algorithms_with_insufficient_noise.cc

```diff
@@ -0,0 +1,187 @@
+      num_datasets, num_samples_per_histogram);
+
+    bool algo_is_dp = tester.Run(); // false = 0, true = 1
+    std::cout << algo_is_dp << std::endl; 
+    return algo_is_dp; // correct outcome is false, e.g., DP premise rejected
+
+  }
+
+  else if (algorithm == "boundedmean") {
+
+    auto algorithm = absl::make_unique<MeanWithInsufficientNoise<double>>(
+      std::log(3),sequence->RangeMin(), sequence->RangeMax(),
+      absl::make_unique<test_utils::SeededLaplaceMechanism::Builder>(), ratio);
+
+    StochasticTester<double> tester(std::move(algorithm), std::move(sequence),
+      num_datasets, num_samples_per_histogram);
+
+    bool algo_is_dp = tester.Run(); // false = 0, true = 1
+    std::cout << algo_is_dp << std::endl; 
+    return algo_is_dp; // correct outcome is false, e.g., DP premise rejected
+  }
+
+  else {
+    std::cout << "The algorithm specified was not recognized." << 
+      "Try count, boundedsum, or boundedmean instead." << std::endl;
+  }
+}
+
+
+void GetTestResults(std::ofstream& datafile, std::string algorithm,
+  double num_datasets, double num_samples_per_histogram, double ratio_min,
+  double ratio_max) {
+
+  double num_tests = 0;
+  double num_tests_passed = 0;
+  double maximum_ratio_passed = 0;
+  auto start = std::chrono::high_resolution_clock::now();
+
+  for (double i=ratio_min; i<=ratio_max; i++) {
+    auto start_test_run = std::chrono::high_resolution_clock::now();
+    double ratio = i/100.0;
+    std::cout << "Now calculating " << algorithm << " algorithm with ratio: " << 
+      ratio << std::endl;
+    bool outcome = RunStochasticTester(algorithm, ratio, num_datasets,
+      num_samples_per_histogram);
+    num_tests++;
+
+    if (outcome == 0) { // correct outcome is false, e.g., DP predicate rejected 
+      num_tests_passed++;
+      maximum_ratio_passed = ratio;
+    }
+
+    auto finish_test_run = std::chrono::high_resolution_clock::now();
+    std::chrono::duration<double> test_run_length =
+      std::chrono::duration_cast<std::chrono::duration<double>>(
+      finish_test_run - start_test_run);
+    datafile << "insufficient_noise," << algorithm << ",0," << outcome << "," <<
+      ratio << "," << num_datasets << "," << num_samples_per_histogram << "," <<
+      test_run_length.count() << "\n";
+  }
+
+  auto end = std::chrono::high_resolution_clock::now();
+  std::chrono::duration<double> time_elapsed =
+    std::chrono::duration_cast<std::chrono::duration<double>>(end - start);
+  std::cout << num_tests_passed << " out of " << num_tests << " passed. " <<
+    "The largest ratio value that passed was " << maximum_ratio_passed << "." <<
+    std::endl;
+    }
+  } // namespace testing
+} // namespace differential_privacy
+
+int main(int argc, char *argv[]) {
+
+  double num_datasets = 15;
+  double num_samples_per_histogram = 1000000;
+  double ratio_min = 80.0;
+  double ratio_max = 99.0;
+  std::ofstream datafile;
```

Comment:
Could we allow these to (optionally) be input at the command-line like we do for the file path of dataFile? (See lines 165-167)

Reply:
I don't see why not!

Reply:
So upon further consideration, I settled on allowing ratio_min, ratio_max, and num_samples_per_histogram to be user-defined and num_datasets remaining immutable. Bc num_datasets is linked to StatisticalTester so has to be exact.